{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ea153251",
   "metadata": {},
   "source": [
    "# Parser for PAS-X Data File\n",
    "This notebook focuses on parsing data from PAS-X software to extract and model/organize high-cell density fermentation data.  \n",
    "\n",
    "PAS-X file contains data about one project which contains several batches of different runs, and each run contains data of several features.   \n",
    "\n",
    "The <b>primary objective</b> is to parse the PAS-X data file with configs files such as parser_config and run_config in order to generate document structures (JSON) for our graph, concretely node documents and edge documents such as: project, batch_cell_cultures, run, strain, species, initial_conditions and process_conditions.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3b96a6a9-cc90-46f7-9a29-ed9e73623d8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Install dependencies\n",
    "# !pip install pandas\n",
    "# !pip install pyarango"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "33ae25ad-ad5c-4bff-957c-1d9a124d7864",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Import modules\n",
    "import os\n",
    "import yaml\n",
    "import json\n",
    "import hashlib\n",
    "from functools import reduce\n",
    "import operator\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "30c3e111-55d5-4282-a0eb-515d579d85d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. Define file paths\n",
    "data_dir = '../data'\n",
    "data_path = os.path.join(data_dir, \"CrashMS_high resolution HD experiment_synth.json\")\n",
    "config_dir = '../config'\n",
    "config_path = os.path.join(config_dir, 'parser_config.yaml')\n",
    "run_config_path = os.path.join(config_dir, 'run_config.yaml')\n",
    "output_dir = '../output'\n",
    "nodes_path = os.path.join(output_dir, \"fermentdb_nodes_full.json\")\n",
    "edges_path = os.path.join(output_dir, \"fermentdb_edges_full.json\")\n",
    "imodulon_files = {'e_coli': {'precise1k': os.path.join(data_dir, 'iM_table.csv')}}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28a5b857",
   "metadata": {},
   "source": [
    "### 1. Load data and config files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "e34617eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['modification_time', 'result_notes', 'tags', 'departments', 'event_names', 'start', 'key_variable', 'editable', 'sites', 'creation_time', 'end', 'workflow', 'created_by', 'manager', 'trending_settings', 'editable_plots', 'country', 'unit_operations', 'name', 'description', 'deletable', 'type', 'batch_phase_names', 'project', 'progress', 'id', 'status', 'batches', 'variables'])"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load PASX data file\n",
    "with open(data_path, 'r') as dbfile:\n",
    "    data = json.load(dbfile)\n",
    "\n",
    "data.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "42640478",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Project': {'_key': 'project',\n",
       "  'id': 'project',\n",
       "  'name': 'name',\n",
       "  'description': 'description',\n",
       "  'progress': 'progress',\n",
       "  'status': 'status',\n",
       "  'deletable': 'deletable',\n",
       "  'editable': 'editable',\n",
       "  'type': 'type',\n",
       "  'creation_time': 'creation_time',\n",
       "  'tags': 'tags',\n",
       "  'modification_time': 'modification_time',\n",
       "  'start': 'start',\n",
       "  'end': 'end'},\n",
       " 'Country': {'_key': 'country', 'id': 'country', 'name': 'country'},\n",
       " 'User': {'_key': 'manager/username',\n",
       "  'id': 'manager/username',\n",
       "  'name': 'manager/first_name',\n",
       "  'surname': 'manager/last_name'},\n",
       " 'Batch_cell_culture': {'_key': 'id', 'id': 'id', 'name': 'name'},\n",
       " 'Run': {'batches': {'id': 'id',\n",
       "   'name': 'name',\n",
       "   'description': 'description',\n",
       "   'creation_time': 'creation_time',\n",
       "   'modification_time': 'modification_time',\n",
       "   'batch_start': 'batch_start',\n",
       "   'batch_end': 'batch_end',\n",
       "   'first_timestamp': 'first_timestamp',\n",
       "   'last_timestamp': 'last_timestamp',\n",
       "   'variables': 'variables',\n",
       "   'phases': 'phases'}}}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load parser_config.yaml\n",
    "with open(config_path, 'r') as cfile:\n",
    "    config = yaml.load(cfile, Loader=yaml.SafeLoader)\n",
    "\n",
    "config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6d0c465e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Run': {'id': 'id',\n",
       "  'Culture Type': 'culture_type',\n",
       "  'Seed': 'seed',\n",
       "  'Sample ID': 'sample_id',\n",
       "  'Experiment': 'experiment',\n",
       "  'Replicate #': 'replicate_number',\n",
       "  'Strain Batch': 'strain_batch',\n",
       "  'Container ID (calculated)': 'container_id',\n",
       "  'Reactor/Plate/Flask Number': 'reactor_plate_flask_number',\n",
       "  'Control?': 'is_control',\n",
       "  'Comments': 'comments',\n",
       "  'Inducer': 'inducer',\n",
       "  'Volume Unit_INDUCTION': 'volume_unit_induction',\n",
       "  'Condition': 'condition'},\n",
       " 'Fermenter': {'Container Type': 'name'},\n",
       " 'Initial_condition': {'SOP': 'SOP',\n",
       "  'Aeration Gas Type': 'aeration_gas_type',\n",
       "  'Aeration Profile': 'aeration_profile',\n",
       "  'Maximum Aeration (slpm)': 'maximum_aeration_slpm',\n",
       "  'Minimum Aeration (slpm)': 'minimum_aeration_slpm',\n",
       "  'Maximum Stirring or Shaking Speed (rpm)': 'maximum_stirring_speed_rpm',\n",
       "  'Minimum Stirring or Shaking Speed (rpm)': 'minimum_stirring_speed_rpm',\n",
       "  'DO Control Setpoint (%)': 'DO_control_setpoint_percentage',\n",
       "  'DO Control Cascade Level 1': 'DO_control_cascade_level1',\n",
       "  'DO Control Cascade Level 2': 'DO_control_cascade_level2',\n",
       "  'DO Control Cascade Level 3': 'DO_control_cascade_level3',\n",
       "  'Starting OD': 'starting_OD',\n",
       "  'Feed #1 Profile': 'feed_1_profile',\n",
       "  'Feed Medium #1': 'feed_medium_1',\n",
       "  'Feed Medium #2': 'feed_medium_2',\n",
       "  'pH Setpoint': 'pH_setpoint',\n",
       "  'pH Control Base Solution': 'ph_control_base_solution',\n",
       "  'pH Control Acid Solution': 'ph_control_acid_solution',\n",
       "  'Temperature Setpoint (C)': 'temperature_setpoint_celsius',\n",
       "  'Main Culture ID': 'main_culture_id',\n",
       "  'Initial Culture VolumeNone': 'initial_volume',\n",
       "  'Culture Volume Unit': 'culture_volume_unit',\n",
       "  'Base Medium': 'base_medium'}}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load run_config.yaml\n",
    "with open(run_config_path, 'r') as cfile:\n",
    "    rconfig = yaml.load(cfile, Loader=yaml.SafeLoader)\n",
    "\n",
    "rconfig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c29be854",
   "metadata": {},
   "source": [
    "### 2. Create node collections dictionary structure\n",
    "Create a dictionary structure for nodes collections with parser_config file and PASX data file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "03a4becc-79a7-4191-a505-decba440eb44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function Definitions\n",
    "\n",
    "def get_from_nested_dict(data_dict:dict, map_list:list):\n",
    "    '''\n",
    "    Extracts value from a nested dictionary given a list of keys (different levels).\n",
    "\n",
    "    parameters:\n",
    "        data_dict (dict): dictionary structure \n",
    "        map_list (list): list of nested keys\n",
    "\n",
    "    return:\n",
    "        nested_value: nested value given the provided list of keys\n",
    "\n",
    "    example:\n",
    "        >>> data_dict = {'a': {'b': {'c': 5}}}\n",
    "        >>> value = get_from_nested_dict(data_dict=data_dict, map_list=['a', 'b', 'c'])\n",
    "        >>> print(value)\n",
    "        5\n",
    "    '''\n",
    "\n",
    "    nested_value = None\n",
    "    try:\n",
    "        nested_value = reduce(operator.getitem, map_list, data_dict)\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    return nested_value\n",
    "\n",
    "def get_hash(key, prefix=\"\"):\n",
    "    '''\n",
    "    Get a hash value for a given key:\n",
    "    Hash a string using SHA-1, before encode key string into bytes using the UTF-8 encoding, as the sha1() expects bytes as input. \n",
    "    Convert the binary hash value into a hexadecimal string and then into an 8-digit integer.\n",
    "    Convert it to a string and add a prefix.\n",
    "\n",
    "    parameters:\n",
    "        key (str): Input string to be hashed.\n",
    "        prefix (str, optional): Prefix to prepend to the hash value. Defaults to \"\".\n",
    "\n",
    "    return:\n",
    "        str: Hash value.\n",
    "\n",
    "    example:\n",
    "        >>> input_key = \"example_key\"\n",
    "        >>> get_hash(input_key, prefix=\"HASH_\")\n",
    "        'HASH_45200d86'\n",
    "    '''\n",
    "    hkey = str(int(hashlib.sha1(key.encode(\"utf-8\")).hexdigest(),16) % (10 ** 8))\n",
    "    hkey = f\"{prefix}{hkey}\"\n",
    "    \n",
    "    return hkey\n",
    "\n",
    "\n",
    "def get_collections_from_pasx(data: dict, config: dict) -> dict:\n",
    "    '''\n",
    "    This function creates a dictionary with the node collections expected\n",
    "    in FermentDB. It requires a dictionary with the data exported from PASX\n",
    "    in json format, and a configuration file that specifies the mapping between\n",
    "    PASX and FermentDB structure. See example in '/config/parser_config.yaml'\n",
    "\n",
    "    parameters:\n",
    "        data (dict): dictionary with the data read from PASX in json format\n",
    "        config (dict): mapping configuration to adapt to FermentDB structure\n",
    "\n",
    "    return:\n",
    "        collections (dict): dictionary with the expected node objects in FermentDB\n",
    "\n",
    "    example:\n",
    "        >>> get_collections_from_pasx(data=pasx_json_dict, config=parser_config.yaml)\n",
    "    '''\n",
    "    collections = {}\n",
    "    for c in config:\n",
    "        collections[c] = {}\n",
    "        # collections[c] = []\n",
    "        for a in config[c]:\n",
    "            if type(config[c][a]) != dict:\n",
    "                key = config[c][a].split('/')\n",
    "                value = get_from_nested_dict(data, key)\n",
    "                if a == '_key':\n",
    "                    value = str(value)\n",
    "                    if c == 'Project':  \n",
    "                        value = get_hash(value, prefix=\"P\")\n",
    "                    elif c == 'Batch_cell_culture':  \n",
    "                        value = get_hash(value, prefix=\"B\")\n",
    "                    elif c == 'User':  \n",
    "                        value = get_hash(value, prefix=\"U\")\n",
    "                    elif c == 'Country':  \n",
    "                        value = get_hash(value, prefix=\"Co\")\n",
    "                collections[c].update({a: value})\n",
    "                # collections[c].append({a: value})\n",
    "            else:\n",
    "                collections[c] = []\n",
    "                for nested_data in data[a]:\n",
    "                    nested_collection = {}\n",
    "                    for nested_a in config[c][a]:\n",
    "                        key = config[c][a][nested_a].split('/')\n",
    "                        value = get_from_nested_dict(nested_data, key)\n",
    "                        if nested_a == '_key':\n",
    "                            value = str(value)\n",
    "                        nested_collection.update({nested_a: value})\n",
    "                    collections[c].append(nested_collection)\n",
    "                    \n",
    "    return collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "848efb7c-d2a1-4022-95a4-8582b7ecb645",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create node collections \n",
    "collections = get_collections_from_pasx(data, config)\n",
    "\n",
    "# print(collections[\"Project\"])\n",
    "# print(collections[\"User\"])\n",
    "# print(collections[\"Country\"])\n",
    "# print(collections[\"Batch_cell_culture\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e863d2a",
   "metadata": {},
   "source": [
    "### 3. Create edges collection and update node collections\n",
    "- Create a dictionary for edges collections.\n",
    "- Extract variables and phases from Run Collection and create node collections: initial_conditions, process_condition, Strain and phase_event. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5f3f4753",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function Definitions\n",
    "\n",
    "def get_unique_json_from_list_of_dicts(d, unique_key='id'):\n",
    "    '''\n",
    "    Get a list of unique dictionaries based on a specified key.\n",
    "    \n",
    "    parameters:\n",
    "        d (list): list of dictionaries.\n",
    "        unique_key (str, optional): Key to determine uniqueness. Default to 'id'.\n",
    "    \n",
    "    return: \n",
    "        list: list of unique dictionaries. \n",
    "    \n",
    "    example:\n",
    "        >>> example_list = [\n",
    "        ...     {\"id\": 1, \"name\": \"Alice\"},\n",
    "        ...     {\"id\": 2, \"name\": \"Bob\"},\n",
    "        ...     {\"id\": 1, \"name\": \"Charlie\"},  # Duplicate id\n",
    "        ...     {\"id\": 3, \"name\": \"Alice\"}     # Duplicate name\n",
    "        ... ]\n",
    "        >>> get_unique_json_from_list_of_dicts(example_list)\n",
    "        #Intermediate step:\n",
    "        # {1: {\"id\": 1, \"name\": \"Alice\"}, 2: {\"id\": 2, \"name\": \"Bob\"}, 3: {\"id\": 3, \"name\": Alice}}.\n",
    "        [{'id': 1, 'name': 'Alice'}, {'id': 2, 'name': 'Bob'}, {'id': 3, 'name': 'Alice'}]\n",
    "    '''\n",
    "    unique_list = list({v[unique_key]:v for v in d}.values())\n",
    "    \n",
    "    return unique_list\n",
    "\n",
    "def get_run_conditions(collections):\n",
    "    '''\n",
    "    Update node collections structure with nodes: initial condition, process conditions, fermenter, and strain.\n",
    "    Generate edges structure with collections: has_initial_condition, has_condition, cultures_strain, and uses_fermenter.  \n",
    "    \n",
    "    parameter:\n",
    "    - collections (dict): A dictionary containing collections of data.\n",
    "\n",
    "    returns:\n",
    "    - edges (dict): A dictionary containing edges data representing the relationships between runs, conditions, and strain.\n",
    "    '''\n",
    "    iconditions_collection = []\n",
    "    pconditions_collection = []\n",
    "    fermenter_collection = []\n",
    "    strain_mapping = {}\n",
    "    fake_name_counter = 1\n",
    "    edges = {'has_initial_condition': {'edges':[],\n",
    "                                        'from_collection': ['Run'],\n",
    "                                        'to_collection': ['Initial_condition']},\n",
    "                'has_condition': {'edges': [],\n",
    "                                'from_collection': ['Run'],\n",
    "                                'to_collection': ['Process_condition']},\n",
    "                'cultures_strain': {'edges': [],\n",
    "                                'from_collection': ['Run'],\n",
    "                                'to_collection': ['Strain']},\n",
    "                'uses_fermenter': {'edges': [],\n",
    "                                'from_collection': ['Run'],\n",
    "                                'to_collection': ['Fermenter']},\n",
    "            }\n",
    "    \n",
    "    for run in collections['Run']:\n",
    "        run[\"_key\"] = run[\"name\"]+\"_\"+str(run['id'])\n",
    "        run[\"_key\"] = get_hash(run[\"_key\"], prefix=\"R\")\n",
    "        for variable in run['variables']:\n",
    "            variable[\"_key\"] = str(variable['name'])\n",
    "            variable[\"_key\"] = get_hash(variable[\"_key\"], prefix=\"C\")\n",
    "            data = variable.pop('data')\n",
    "            timestamps = variable.pop('timestamps')\n",
    "            unit = variable.pop('unit')\n",
    "            _ = variable.pop('categorical_data')\n",
    "            _ = variable.pop('raw_data')\n",
    "            _ = variable.pop('datetime_data')\n",
    "            _ = variable.pop('errors')\n",
    "            \n",
    "            if variable['name'] in rconfig['Run']:\n",
    "                if type(data) == list:\n",
    "                    data = data[0][0]\n",
    "                if variable['name'] == \"Strain Batch\":\n",
    "                    # Create fake strain\n",
    "                    original_strain = '_'.join(data.split('-')[:1])\n",
    "                    if original_strain not in strain_mapping:\n",
    "                        fake_name = f\"Strain{fake_name_counter}\"\n",
    "                        strain_mapping[original_strain] = fake_name\n",
    "                        fake_name_counter += 1\n",
    "                    else:\n",
    "                        fake_name = strain_mapping[original_strain]\n",
    "                    strain = fake_name\n",
    "                    strain_key = get_hash(strain, prefix=\"ST\")\n",
    "                    if 'Strain' not in collections: # Create Strain collection\n",
    "                        collections['Strain'] = []\n",
    "                    collections['Strain'].append({'_key': strain_key,\n",
    "                                                'name': strain,\n",
    "                                                'rank': 'strain'})\n",
    "                    edges['cultures_strain']['edges'].append({'_from': f\"Run/{run['_key']}\",\n",
    "                                                            '_to': f'Strain/{strain_key}',\n",
    "                                                            'strain_batch': data})\n",
    "                run.update({rconfig['Run'][variable['name']]: data}) # Add to edge cultures strain and delete?\n",
    "            elif variable['name'] in rconfig['Fermenter']:\n",
    "                variable[\"_key\"] = get_hash(data, prefix=\"F\")\n",
    "                fermenter_collection.append({'_key': variable[\"_key\"],\n",
    "                                             'name': data})\n",
    "                edges['uses_fermenter']['edges'].append({'_from': f\"Run/{run['_key']}\",\n",
    "                           '_to': f\"Fermenter/{variable['_key']}\"})\n",
    "            elif variable['name'] in rconfig['Initial_condition']:\n",
    "                iconditions_collection.append(variable)\n",
    "                edges['has_initial_condition']['edges'].append({'_from': f\"Run/{run['_key']}\",\n",
    "                           '_to': f\"Initial_condition/{variable['_key']}\",\n",
    "                           'data': data,\n",
    "                           'unit': unit})\n",
    "            elif '_RNAseq' in variable['name']:\n",
    "                if not all(v == 0 for v in data):\n",
    "                    variable['name'] = ' '.join(variable['name'].replace('_RNAseq', '').split('_'))\n",
    "                    variable[\"_key\"] = get_hash(variable[\"name\"], prefix=\"iM\")\n",
    "                    edges['has_measured_imodulon']['edges'].append({'_from': f'Run/{run['_key']}',\n",
    "                                                           '_to': f'iModulon/{variable['_key']}',\n",
    "                                                           'data': data,\n",
    "                                                           'timestamps': timestamps})\n",
    "            else:\n",
    "                pconditions_collection.append(variable)\n",
    "                if not all(v == 0 for v in data):\n",
    "                    edges['has_condition']['edges'].append({'_from': f\"Run/{run['_key']}\",\n",
    "                            '_to': f\"Process_condition/{variable['_key']}\",\n",
    "                            'data': data,\n",
    "                            'unit': unit,\n",
    "                            'timestamps': timestamps})\n",
    "        del run['variables']\n",
    "    \n",
    "    iconditions_collection = get_unique_json_from_list_of_dicts(d=iconditions_collection, \n",
    "                                                                    unique_key='_key')\n",
    "    pconditions_collection = get_unique_json_from_list_of_dicts(d=pconditions_collection, \n",
    "                                                                    unique_key='_key')\n",
    "    fermenter_collection = get_unique_json_from_list_of_dicts(d=fermenter_collection, \n",
    "                                                                    unique_key='_key')\n",
    "    collections['Strain'] = get_unique_json_from_list_of_dicts(d=collections['Strain'], \n",
    "                                                                    unique_key='_key')\n",
    "    collections['Initial_condition'] = iconditions_collection\n",
    "    collections['Process_condition'] = pconditions_collection\n",
    "    collections['Fermenter'] = fermenter_collection\n",
    "    \n",
    "    return edges\n",
    "\n",
    "def get_run_phases(collections):\n",
    "    '''\n",
    "    Retrieves phases data from Run collection and create Phase_event node collection and has_phase edge collection.  \n",
    "\n",
    "    parameter:\n",
    "    - collections (dict): A dictionary containing collections of data.\n",
    "\n",
    "    returns:\n",
    "    - dict: A dictionary containing edges data representing the relationships between runs and phases.\n",
    "    '''\n",
    "    phases_collection = []\n",
    "    edges = {'has_phase': {'edges':[],\n",
    "                           'from_collection': ['Run'],\n",
    "                           'to_collection':['Phase_event']}}\n",
    "\n",
    "    for run in collections['Run']:\n",
    "        for phase in run['phases']:\n",
    "            phase[\"_key\"] = phase['name']\n",
    "            phase[\"_key\"] = get_hash(phase[\"_key\"], prefix=\"PH\")\n",
    "            attributes = {\"event_start\": phase.pop(\"start\"),\n",
    "                          \"event_end\": phase.pop(\"end\"),\n",
    "                          \"comment\": phase.pop(\"comment\"),\n",
    "                          \"created_by\": phase.pop(\"created_by\"),\n",
    "                          \"start\": phase.pop(\"relative_start\"),\n",
    "                          \"end\": phase.pop(\"relative_end\")}\n",
    "            attributes.update({'_from': f\"Run/{run['_key']}\",\n",
    "                           '_to': f\"Phase_event/{phase['_key']}\"})\n",
    "            phases_collection.append(phase)\n",
    "            edges['has_phase']['edges'].append(attributes)\n",
    "            phases_collection = get_unique_json_from_list_of_dicts(d=phases_collection, \n",
    "                                                                    unique_key='_key')\n",
    "        del run['phases']\n",
    "    \n",
    "    collections['Phase_event'] = phases_collection\n",
    "\n",
    "    return edges\n",
    "    \n",
    "\n",
    "def generate_project_edges(collections, edges):\n",
    "    '''\n",
    "    Generate edges between collections: has_batch, created_by, and has_run; between nodes: user, project and batch_cell_culture nodes collections.\n",
    "    \n",
    "    parameter: \n",
    "        collections (dict): A dictionary containing collections as keys and their corresponding documents as values.\n",
    "        edges(dict): A dictionary representing edges between collections.\n",
    "    \n",
    "    returns:\n",
    "        None: This function updates the 'edges' dictionary in place.\n",
    "    \n",
    "    example:\n",
    "        collections = {\n",
    "        'Project': {'_key': 'project_key', 'creation_time': '2024-05-01T12:00:00Z'},\n",
    "        'User': {'_key': 'user_key'},\n",
    "        'Batch_cell_culture': {'_key': 'batch_key'},\n",
    "        'Run': [{'_key': 'run_key1'}, {'_key': 'run_key2'}, ...]\n",
    "        }\n",
    "        edges = {}\n",
    "        generate_project_edges(collections, edges)\n",
    "        # edges will be updated with edges between collections.\n",
    "    '''\n",
    "   \n",
    "    # edges.update({'has_batch': {'edges':[], \n",
    "    #                             'from_collection': ['Project'],\n",
    "    #                             'to_collection':['Batch_cell_culture']},\n",
    "    #             'created_by': {'edges':[],\n",
    "    #                             'from_collection': ['Project'],\n",
    "    #                             'to_collection':['User']},\n",
    "    #             'has_run': {'edges':[], \n",
    "    #                         'from_collection': ['Batch_cell_culture'],\n",
    "    #                         'to_collection':['Run']}\n",
    "    #                             })\n",
    "    \n",
    "    # for user in collections['User']:\n",
    "    #     for batch in collections['Batch_cell_culture']:\n",
    "    #         for project in collections['Project']:\n",
    "    #             if '_key' in project:\n",
    "    #                 edges['has_batch']['edges'].append({'_from': f\"Project/{project['_key']}\",\n",
    "    #                                                     '_to': f\"Batch_cell_culture/{batch['_key']}\"})\n",
    "    #                 edges['created_by']['edges'].append({'_from': f\"Project/{project['_key']}\",\n",
    "    #                                                     '_to': f\"User/{user['_key']}\"})\n",
    "    project = collections['Project']\n",
    "    user = collections['User']\n",
    "    batch = collections['Batch_cell_culture']\n",
    "\n",
    "    edges.update({'has_batch': {'edges':[{'_from': f\"Project/{project['_key']}\",\n",
    "                                          '_to': f\"Batch_cell_culture/{batch['_key']}\"}], \n",
    "                                'from_collection': ['Project'],\n",
    "                                'to_collection':['Batch_cell_culture']},\n",
    "                  'created_by': {'edges':[{'_from': f\"Project/{project['_key']}\",\n",
    "                                          '_to': f\"User/{user['_key']}\",\n",
    "                                          'creation_date': project['creation_time']}],\n",
    "                                'from_collection': ['Project'],\n",
    "                                'to_collection':['User']}})\n",
    "    edges.update({'has_run': {'edges':[], \n",
    "                              'from_collection': ['Batch_cell_culture'],\n",
    "                                'to_collection':['Run']}})\n",
    "    for run in collections['Run']:\n",
    "        edges['has_run']['edges'].append({'_from': f\"Batch_cell_culture/{batch['_key']}\",\n",
    "                                          '_to': f\"Run/{run['_key']}\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e3667915",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create edges with conditions collections, phases collections and project edges\n",
    "\n",
    "edges = get_run_conditions(collections)\n",
    "edges.update(get_run_phases(collections))\n",
    "generate_project_edges(collections, edges)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3ab9977",
   "metadata": {},
   "source": [
    "### 4. Create species collection and edge collection.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "0055c6bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a Species collection\n",
    "if 'Species' not in collections:\n",
    "    species_key = get_hash(\"Escherichia coli\", prefix=\"SP\")\n",
    "    collections['Species'] = [{'_key': species_key,\n",
    "                            'name': 'Escherichia coli',\n",
    "                            'rank': 'species'}]\n",
    "\n",
    "# Create edge 'belongs_to'\n",
    "if 'belongs_to' not in edges:\n",
    "    edges['belongs_to'] = {\n",
    "        'edges': [],\n",
    "        'from_collection': ['Strain'],\n",
    "        'to_collection': ['Species']\n",
    "    }\n",
    "\n",
    "for strain in collections['Strain']:\n",
    "    strain_key = strain['_key']\n",
    "    edges['belongs_to']['edges'].append({\n",
    "        '_from': f\"Strain/{strain_key}\",\n",
    "        '_to': f'Species/{species_key}'\n",
    "    }) \n",
    "                                "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ec2777b",
   "metadata": {},
   "source": [
    "### 5. Create institution node collection and edges collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "39c25ea2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate an Institution node collection with a document at its edges\n",
    "if 'Institution' not in collections:\n",
    "        institution_key = get_hash(\"NNFCB\", prefix=\"I\")\n",
    "        collections['Institution'] = [{'_key': institution_key,\n",
    "                                'name': 'NNFCB - Novo Nordisk Foundation Center for Biosustainability (DTU Biosustain)',\n",
    "                                'address': 'Building 220, Kemitorvet. 2800 Kgs. Lyngby',\n",
    "                                'email': 'biosustain@biosustain.dtu.dk',\n",
    "                                'phone_number':'+45 45 25 80 00'\n",
    "                                }]\n",
    "\n",
    "project = collections['Project']\n",
    "user = collections['User']\n",
    "country = collections['Country']\n",
    "\n",
    "edges.update({'created_at': {'edges':[{'_from': f\"Project/{project['_key']}\",\n",
    "                                          '_to': f\"Institution/{institution_key}\"}], \n",
    "                                'from_collection': ['Project'],\n",
    "                                'to_collection':['Institution']},\n",
    "                'works_at': {'edges':[{'_from': f\"User/{user['_key']}\",\n",
    "                                          '_to': f\"Institution/{institution_key}\"}],\n",
    "                                'from_collection': ['User'],\n",
    "                                'to_collection':['Institution']},\n",
    "                'from': {'edges':[{'_from': f\"Institution/{institution_key}\",\n",
    "                                          '_to': f\"Country/{country['_key']}\",\n",
    "                                          'creation_date': project['creation_time']}],\n",
    "                                'from_collection': ['Institution'],\n",
    "                                'to_collection':['Country']}})\n",
    "\n",
    "# ## Create 'created_at' edge\n",
    "# if 'created_at' not in edges:\n",
    "#         edges['created_at'] = {'edges': [],\n",
    "#                                 'from_collection': ['Project'],\n",
    "#                                 'to_collection': ['Institution']}\n",
    "\n",
    "# for project in collections['Project']:\n",
    "#         print(f\"Project: {project}\")\n",
    "#         project_key = int(project[\"_key\"])\n",
    "#         edges['created_at']['edges'].append({\n",
    "#                 '_from': f\"Project/{project_key}\",\n",
    "#                 '_to': f\"Institution/{institution_key}\"\n",
    "#         })\n",
    "\n",
    "# ## Create 'works_at' edge\n",
    "# if 'works_at' not in edges:\n",
    "#         edges['works_at'] = {'edges':[], \n",
    "#                                 'from_collection': ['User'],\n",
    "#                                 'to_collection': ['Institution']}\n",
    "\n",
    "# for user in collections['User']:\n",
    "#         user_key = int(user[\"_key\"])\n",
    "#         edges['works_at']['edges'].append({\n",
    "#                 '_from': f\"User/{user_key}\",\n",
    "#                 '_to': f\"Institution/{institution_key}\"\n",
    "#         })\n",
    "\n",
    "\n",
    "\n",
    "# ## Create 'from' edge\n",
    "# if 'from' not in edges:\n",
    "#         edges['from'] = {'edges': [],\n",
    "#                                 'from_collection': ['Institution'],\n",
    "#                                 'to_collection': ['Country']}\n",
    "\n",
    "# for country in collections['Country']:\n",
    "#         country_key = int(country[\"_key\"])\n",
    "#         edges['from']['edges'].append({\n",
    "#                 '_from': f\"Institution/{institution_key}\",\n",
    "#                 '_to': f\"Country/{country_key}\"\n",
    "#         })\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10858d76",
   "metadata": {},
   "source": [
    "### 6. Create iModulon node and edges collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "4bfdf0ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Define iModulon function \n",
    "def get_imodulon_collection(organism, dataset, table_path):\n",
    "    '''\n",
    "    Process iModulon data to generate a collection for a given organism and dataset. \n",
    "    - Reads data from a CSV file into a pandas DataFrame. \n",
    "    - Adds additional fields such as a key column from hashing name column, and creates linkout column by constructing URLs to generate individual links for each iModulon entry.\n",
    "    - Processes dataset by handling misisng values and dropping the 'k' column as it is no loger needed (values areincorporated into the linkout column) \n",
    "\n",
    "    parameters:\n",
    "        organism (str): The name of the organism (e.g., 'e_coli').\n",
    "        dataset (str): The name of the dataset (e.g., 'precise1k').\n",
    "        table_path (str): The file path to the CSV file containing the iModulon data (e.g., '/path/to/iM_table.csv').\n",
    "\n",
    "    return:\n",
    "        data_dict: A list of dictionaries representing processed iModulon data\n",
    "    '''\n",
    "    imodulon_link = f'https://imodulondb.org/iModulon.html?organism={organism}&dataset={dataset}&k='\n",
    "    data = pd.read_csv(table_path, sep=',', header=0)\n",
    "    data['_key'] = data['name'].apply(lambda n: get_hash(n, prefix=\"iM\"))\n",
    "    data['linkout'] = data['k'].apply(lambda k: imodulon_link+str(k))\n",
    "    data = data.fillna('NaN').drop('k', axis=1)\n",
    "    data_dict = data.to_dict(orient='records')\n",
    "    return data_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f28be55b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'name': 'Sugar Diacid', 'regulator_readable': 'CdaR', 'function': 'Transport and catabolism of carbohydrate diacids (galacatarate and glucarate)', 'category': 'Uncharacterized', 'n_genes': 14, 'exp_var': 0.0024127485650338, 'precision': 0.6428571428571429, 'recall': 1.0, 'category_num': 0, '_key': 'iM21437567', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=0'}, {'name': 'Translation', 'regulator_readable': 'DksA', 'function': 'Ribosome biogenesis', 'category': 'Uncharacterized', 'n_genes': 53, 'exp_var': 0.0074692449365292, 'precision': 0.8301886792452831, 'recall': 0.3928571428571428, 'category_num': 1, '_key': 'iM57743932', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=1'}, {'name': 'ygdT', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 3, 'exp_var': 0.000344428784103, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 2, '_key': 'iM6684080', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=2'}, {'name': 'OxyR', 'regulator_readable': 'OxyR', 'function': 'Oxidative stress response, especially peroxide', 'category': 'Uncharacterized', 'n_genes': 38, 'exp_var': 0.0017473899530303, 'precision': 0.4210526315789473, 'recall': 0.5517241379310345, 'category_num': 3, '_key': 'iM27743110', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=3'}, {'name': 'FlhDC-2', 'regulator_readable': 'FlhDC', 'function': 'Flagellum biosynthesis and export', 'category': 'Uncharacterized', 'n_genes': 43, 'exp_var': 0.0300211202151899, 'precision': 0.8604651162790697, 'recall': 0.4567901234567901, 'category_num': 4, '_key': 'iM50625738', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=4'}, {'name': 'Osmoprotectant', 'regulator_readable': 'RpoS and ppGpp and IHF and H-NS', 'function': 'Transport of osmoprotectant glycine betaine', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0041890249079538, 'precision': 0.75, 'recall': 0.375, 'category_num': 5, '_key': 'iM83902138', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=5'}, {'name': 'Glutamine', 'regulator_readable': 'NtrC', 'function': 'Glutamine transport and synthesis', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0051219605732515, 'precision': 0.5555555555555556, 'recall': 0.2941176470588235, 'category_num': 6, '_key': 'iM14545735', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=6'}, {'name': 'Sialic Acid', 'regulator_readable': 'Crp and NanR', 'function': 'N-acetyl-neuraminic acid (or sialic acid) transport and metabolism', 'category': 'Uncharacterized', 'n_genes': 13, 'exp_var': 0.0017406664972557, 'precision': 0.6153846153846154, 'recall': 1.0, 'category_num': 7, '_key': 'iM78676814', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=7'}, {'name': 'yqhI', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 2, 'exp_var': 0.0015155493442772, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 8, '_key': 'iM92340874', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=8'}, {'name': 'ROS TALE Amp', 'regulator_readable': 'NaN', 'function': 'Amplification of DLP12/IS3', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0040496685796378, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 9, '_key': 'iM15099127', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=9'}, {'name': 'Nucleotide Stress', 'regulator_readable': 'NaN', 'function': 'DNA repair; stringent response signaling via ppGpp; ribosomal biogenesis and stability', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.0059256313238668, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 10, '_key': 'iM25757213', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=10'}, {'name': 'Histidine', 'regulator_readable': 'His-tRNA and DksA', 'function': 'Histidine biosynthesis', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0024112233799451, 'precision': 0.8888888888888888, 'recall': 1.0, 'category_num': 11, '_key': 'iM73175566', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=11'}, {'name': 'UC-4', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 39, 'exp_var': 0.0031955309222377, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 12, '_key': 'iM40219574', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=12'}, {'name': 'RpoH', 'regulator_readable': 'RpoH', 'function': 'Global heat shock response during log-phase; response to high temperature and/or misfolded proteins', 'category': 'Uncharacterized', 'n_genes': 41, 'exp_var': 0.0039865130407905, 'precision': 0.8292682926829268, 'recall': 0.2428571428571428, 'category_num': 13, '_key': 'iM30098093', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=13'}, {'name': 'FucR/AllR/AraC', 'regulator_readable': 'FucR or AllR or AraC', 'function': 'Allantoin/glycolate/glyoxylate/arabinose/fucose transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 20, 'exp_var': 0.0011734254283444, 'precision': 1.0, 'recall': 0.6666666666666666, 'category_num': 14, '_key': 'iM55048699', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=14'}, {'name': 'ldrC', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0013875276267905, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 15, '_key': 'iM66670137', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=15'}, {'name': 'Ferric Citrate', 'regulator_readable': 'PdhR and HypT and Nac and Crp and Fur and FecI', 'function': 'Ferric citrate transporter; iron uptake under intracellular limitation', 'category': 'Uncharacterized', 'n_genes': 7, 'exp_var': 0.0009606918070365, 'precision': 0.7142857142857143, 'recall': 1.0, 'category_num': 16, '_key': 'iM42811358', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=16'}, {'name': 'Zinc-1', 'regulator_readable': 'Zur or ZntR or CueR', 'function': 'Zinc sensing and uptake; copper sensing and export', 'category': 'Uncharacterized', 'n_genes': 14, 'exp_var': 0.0099165914402776, 'precision': 0.6428571428571429, 'recall': 0.6428571428571429, 'category_num': 17, '_key': 'iM7953674', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=17'}, {'name': 'Methionine', 'regulator_readable': 'MetJ', 'function': 'Methionine transport and biosynthesis', 'category': 'Uncharacterized', 'n_genes': 21, 'exp_var': 0.0066897715297598, 'precision': 0.5238095238095238, 'recall': 0.7333333333333333, 'category_num': 18, '_key': 'iM48561181', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=18'}, {'name': 'Sorbitol', 'regulator_readable': 'SrlR and GutM and Crp and H-NS', 'function': 'Sorbitol transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0026645231421454, 'precision': 1.0, 'recall': 0.8571428571428571, 'category_num': 19, '_key': 'iM73448232', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=19'}, {'name': 'Glycerol', 'regulator_readable': 'Crp and GlpR', 'function': 'Glycerol/glycerol-3-phosphate transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 11, 'exp_var': 0.0072307588112071, 'precision': 0.8181818181818182, 'recall': 1.0, 'category_num': 20, '_key': 'iM98631394', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=20'}, {'name': 'Nucleoside', 'regulator_readable': 'Crp and CytR', 'function': 'Transport and utilization of ribonucleosides and deoxyribonucleosides', 'category': 'Uncharacterized', 'n_genes': 11, 'exp_var': 0.0013260197751645, 'precision': 0.5454545454545454, 'recall': 0.5454545454545454, 'category_num': 21, '_key': 'iM53113789', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=21'}, {'name': 'SoxS', 'regulator_readable': 'SoxS', 'function': 'Superoxide/nitric oxide sensing and removal, protection from organic solvents and antibiotics', 'category': 'Uncharacterized', 'n_genes': 117, 'exp_var': 0.0041681695386095, 'precision': 0.2735042735042735, 'recall': 0.7619047619047619, 'category_num': 22, '_key': 'iM63194991', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=22'}, {'name': 'thrA KO', 'regulator_readable': 'NaN', 'function': 'thrA KO and compensatory transcriptional changes', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0012832693113496, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 23, '_key': 'iM42769539', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=23'}, {'name': 'YjfJ', 'regulator_readable': 'YjfJ', 'function': 'Putative heat responsive regulon', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0021996344850857, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 24, '_key': 'iM3201129', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=24'}, {'name': 'Lrp', 'regulator_readable': 'Lrp', 'function': 'Global regulation of amino acid biosynthesis/catabolism, stationary phase, nutrient transport, pili synthesis, 1-carbon metabolism', 'category': 'Uncharacterized', 'n_genes': 42, 'exp_var': 0.0110477301191561, 'precision': 0.4285714285714285, 'recall': 0.2608695652173913, 'category_num': 25, '_key': 'iM52239795', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=25'}, {'name': 'baeR KO', 'regulator_readable': 'NaN', 'function': 'baeR KO', 'category': 'Uncharacterized', 'n_genes': 54, 'exp_var': 0.0016671227829292, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 26, '_key': 'iM13245910', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=26'}, {'name': 'ydaG', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0015127926125303, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 27, '_key': 'iM49879012', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=27'}, {'name': 'ldrA', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0005526171253633, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 28, '_key': 'iM68885151', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=28'}, {'name': 'Quorum Sensing', 'regulator_readable': 'Crp and LsrR', 'function': 'Quorum sensing via autoinducer 2, biofilm architecture, foreign DNA stress response; lsrR KO', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.0023016299507784, 'precision': 0.9, 'recall': 1.0, 'category_num': 29, '_key': 'iM29439831', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=29'}, {'name': 'IS Elements-2', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 49, 'exp_var': 0.0048016613671348, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 30, '_key': 'iM46886149', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=30'}, {'name': 'Nickel', 'regulator_readable': 'RpoN and NarL and NikR and Fnr', 'function': 'Nickel sensing and uptake', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.00316565882443, 'precision': 1.0, 'recall': 0.8333333333333334, 'category_num': 31, '_key': 'iM72719690', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=31'}, {'name': 'CpxR', 'regulator_readable': 'CpxR', 'function': 'Inner membrane damage sensing and repair via protein folding/degradation, chemotaxis, drug efflux, DNA repair, motility', 'category': 'Uncharacterized', 'n_genes': 53, 'exp_var': 0.0021876125954307, 'precision': 0.3584905660377358, 'recall': 0.3877551020408163, 'category_num': 32, '_key': 'iM23290394', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=32'}, {'name': 'Curli-2', 'regulator_readable': 'RpoS and FliZ and CpxR and CsgD and BtsR', 'function': 'Curli secretion and assembly', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0018032410607966, 'precision': 0.6, 'recall': 0.4285714285714285, 'category_num': 33, '_key': 'iM9462575', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=33'}, {'name': 'Acid Shock RNA', 'regulator_readable': 'NaN', 'function': 'Very low pH-responsive acid shock polypeptide', 'category': 'Uncharacterized', 'n_genes': 3, 'exp_var': 0.0014709159579372, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 34, '_key': 'iM20970501', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=34'}, {'name': 'FucR/ExuR', 'regulator_readable': 'FucR or ExuR', 'function': 'Fucose/galacturonate/glucuronate transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 14, 'exp_var': 0.0018713021866049, 'precision': 0.7857142857142857, 'recall': 0.6875, 'category_num': 35, '_key': 'iM88597141', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=35'}, {'name': 'cydB/appC KO', 'regulator_readable': 'NaN', 'function': 'cydB/appC KO and compensatory regulatory changes', 'category': 'Uncharacterized', 'n_genes': 11, 'exp_var': 0.0011857573787881, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 36, '_key': 'iM88224258', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=36'}, {'name': 'ymdG', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0004899636080474, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 37, '_key': 'iM51053905', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=37'}, {'name': 'EvgA', 'regulator_readable': 'EvgA', 'function': 'Glutamate-dependent acid response, drug resistance, alkali metals response', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.0019328254419214, 'precision': 0.8, 'recall': 0.7058823529411765, 'category_num': 38, '_key': 'iM74266519', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=38'}, {'name': 'Xylose', 'regulator_readable': 'Crp and XylR', 'function': 'd-Xylose transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.0019821871677557, 'precision': 0.4666666666666667, 'recall': 0.875, 'category_num': 39, '_key': 'iM39497951', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=39'}, {'name': 'IS1', 'regulator_readable': 'NaN', 'function': 'IS1 transcripts', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0017638403169903, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 40, '_key': 'iM94332274', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=40'}, {'name': 'pgi KO', 'regulator_readable': 'NaN', 'function': 'pgi KO', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0010002038218784, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 41, '_key': 'iM62228124', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=41'}, {'name': 'Biotin', 'regulator_readable': 'BirA', 'function': 'Biotin biosynthesis', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.0032899000451902, 'precision': 0.625, 'recall': 1.0, 'category_num': 42, '_key': 'iM20417707', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=42'}, {'name': 'N Reduction', 'regulator_readable': 'FlhDC and ModE and NarP and NarL and Fnr', 'function': 'Electron transfer via nitrate reductases (periplasmic and A), cytochrome c', 'category': 'Uncharacterized', 'n_genes': 24, 'exp_var': 0.0049662816437708, 'precision': 0.625, 'recall': 1.0, 'category_num': 43, '_key': 'iM69026461', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=43'}, {'name': 'FDH-N/NRA', 'regulator_readable': 'NarL and Fnr', 'function': 'Anaerobic growth on nitrate; formate-to-nitrate electron transfer via formate dehydrogenase N and nitrate reductase A', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.0034839038834835, 'precision': 0.7333333333333333, 'recall': 0.2037037037037037, 'category_num': 44, '_key': 'iM22682782', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=44'}, {'name': 'Resistance', 'regulator_readable': 'NaN', 'function': 'Multidrug resistance; stress response to harmful compounds/antibiotics', 'category': 'Uncharacterized', 'n_genes': 58, 'exp_var': 0.0041463328835988, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 45, '_key': 'iM22089225', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=45'}, {'name': 'PAL 10 Del', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 108, 'exp_var': 0.0052100640244991, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 46, '_key': 'iM33327388', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=46'}, {'name': 'Biofilm', 'regulator_readable': 'RpoS and ppGpp and BluR and MarA', 'function': 'Biofilm formation, acid resistance, cold shock response', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0024403327678284, 'precision': 0.8, 'recall': 1.0, 'category_num': 47, '_key': 'iM89631000', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=47'}, {'name': 'Cytochrome c', 'regulator_readable': 'FlhDC and IHF and NsrR and NarL and NarP and Fis and Fnr', 'function': 'Cytochrome c-552 and putative cytochrome c-related proteins', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0032992208849464, 'precision': 1.0, 'recall': 0.8571428571428571, 'category_num': 48, '_key': 'iM35214332', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=48'}, {'name': 'BasR', 'regulator_readable': 'BasR', 'function': 'Lipid A modification and polymixin resistance in the presence of excess iron', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.0013302764357293, 'precision': 0.6, 'recall': 0.4090909090909091, 'category_num': 49, '_key': 'iM14781332', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=49'}, {'name': 'Sulfoquinovose', 'regulator_readable': 'CsqR', 'function': 'Sulfoquinovose transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.0011743640365809, 'precision': 0.5, 'recall': 1.0, 'category_num': 50, '_key': 'iM43639060', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=50'}, {'name': 'RpoS', 'regulator_readable': 'RpoS', 'function': 'Master regulator of global stress response', 'category': 'Uncharacterized', 'n_genes': 122, 'exp_var': 0.0391278054178326, 'precision': 0.3934426229508196, 'recall': 0.1476923076923077, 'category_num': 51, '_key': 'iM10588154', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=51'}, {'name': 'lsrB KO', 'regulator_readable': 'NaN', 'function': 'lsrB KO and compensatory regulatory changes', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.0003099282696249, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 52, '_key': 'iM6152804', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=52'}, {'name': 'UC-1', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 27, 'exp_var': 0.0046254928156179, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 53, '_key': 'iM31480000', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=53'}, {'name': 'Fur-1', 'regulator_readable': 'Fur', 'function': 'Iron sensing and uptake, enterobactin biosynthesis, iron-sulfur cluster activation, iron sequestration', 'category': 'Uncharacterized', 'n_genes': 32, 'exp_var': 0.0095775881233364, 'precision': 0.875, 'recall': 0.2204724409448819, 'category_num': 54, '_key': 'iM30622371', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=54'}, {'name': 'ykgR', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0012508412285462, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 55, '_key': 'iM76788890', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=55'}, {'name': 'Sulfate', 'regulator_readable': 'Cbl and CysB', 'function': 'Cysteine transport and biosynthesis; aliphatic sulfonate utilization and homeostatic response to sulfate starvation', 'category': 'Uncharacterized', 'n_genes': 17, 'exp_var': 0.0012151632798734, 'precision': 0.5294117647058824, 'recall': 1.0, 'category_num': 56, '_key': 'iM67826341', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=56'}, {'name': 'ypaB', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0008675182021106, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 57, '_key': 'iM9838494', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=57'}, {'name': 'Fur-2', 'regulator_readable': 'Fur', 'function': 'Iron sensing and uptake, enterobactin biosynthesis, iron-sulfur cluster activation, iron sequestration', 'category': 'Uncharacterized', 'n_genes': 44, 'exp_var': 0.0059071002772245, 'precision': 0.6818181818181818, 'recall': 0.2362204724409448, 'category_num': 58, '_key': 'iM17830048', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=58'}, {'name': 'Magnesium', 'regulator_readable': 'PhoP', 'function': 'Magnesium sensing and transport; magnesium-limitation response', 'category': 'Uncharacterized', 'n_genes': 21, 'exp_var': 0.0013458395093458, 'precision': 0.3809523809523809, 'recall': 0.4, 'category_num': 59, '_key': 'iM82581680', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=59'}, {'name': 'Nitrate/Nitrite', 'regulator_readable': 'NarL and Fnr', 'function': 'Nitrate/nitrite transport and reduction', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.000791493963314, 'precision': 0.8888888888888888, 'recall': 0.1481481481481481, 'category_num': 60, '_key': 'iM59546174', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=60'}, {'name': 'glrR KO', 'regulator_readable': 'NaN', 'function': 'glrR (aka qseF) KO; xylose transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 7, 'exp_var': 0.0003215195076921, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 61, '_key': 'iM64178751', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=61'}, {'name': 'Fnr-1', 'regulator_readable': 'Fnr', 'function': 'Transition from aerobic to anaerobic growth; anaerobic reductase and hydrogenase activity', 'category': 'Uncharacterized', 'n_genes': 46, 'exp_var': 0.0133677234255225, 'precision': 0.6304347826086957, 'recall': 0.0979729729729729, 'category_num': 62, '_key': 'iM39721255', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=62'}, {'name': 'Pyrimidine', 'regulator_readable': 'PurR or UTP', 'function': 'Pyrimidine transport and biosynthesis', 'category': 'Uncharacterized', 'n_genes': 13, 'exp_var': 0.0178792763166786, 'precision': 0.6923076923076923, 'recall': 0.25, 'category_num': 63, '_key': 'iM54707639', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=63'}, {'name': 'Zinc-2', 'regulator_readable': 'ZraR', 'function': 'Tolerance to high zinc concentration', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0008267540018734, 'precision': 1.0, 'recall': 0.3333333333333333, 'category_num': 64, '_key': 'iM51260118', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=64'}, {'name': 'NRZ', 'regulator_readable': 'OmpR', 'function': 'Nitrate reduction under microaerobic condition; aerobic to anaerobic transition', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0047141311728475, 'precision': 0.8333333333333334, 'recall': 0.1515151515151515, 'category_num': 65, '_key': 'iM50224720', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=65'}, {'name': 'nquinone ALE 3', 'regulator_readable': 'NaN', 'function': 'ubiC KO and associated adaptive regulatory changes', 'category': 'Uncharacterized', 'n_genes': 26, 'exp_var': 0.0017680288159783, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 66, '_key': 'iM38117789', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=66'}, {'name': 'RpoE', 'regulator_readable': 'RpoE', 'function': 'Heat shock response; response to high temperature and/or misfolded membrane/periplasmic proteins', 'category': 'Uncharacterized', 'n_genes': 13, 'exp_var': 0.0018326937702909, 'precision': 0.9230769230769232, 'recall': 0.096, 'category_num': 67, '_key': 'iM60332696', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=67'}, {'name': 'Cellulose', 'regulator_readable': 'NaN', 'function': 'Cellulose biosynthesis; biofilm formation; membrane/periplasmic proteins', 'category': 'Uncharacterized', 'n_genes': 95, 'exp_var': 0.008280808853032, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 68, '_key': 'iM55051605', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=68'}, {'name': 'Phosphate-1', 'regulator_readable': 'PhoB', 'function': 'Inorganic phosphate sensing, transport, and production', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.0008951019590812, 'precision': 1.0, 'recall': 0.1702127659574468, 'category_num': 69, '_key': 'iM76702750', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=69'}, {'name': 'Salicylic Acid', 'regulator_readable': 'NaN', 'function': 'Salicylic acid efflux', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0010648013876597, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 70, '_key': 'iM2408758', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=70'}, {'name': 'ymgI', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0009229928906844, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 71, '_key': 'iM592274', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=71'}, {'name': 'Galactose', 'regulator_readable': 'RpoS and GalR and GalS and Crp', 'function': 'Galactose transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.0015237527821455, 'precision': 0.7, 'recall': 1.0, 'category_num': 72, '_key': 'iM82253150', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=72'}, {'name': 'trxC Del', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 3, 'exp_var': 0.0010618121528696, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 73, '_key': 'iM51367832', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=73'}, {'name': 'ArcA', 'regulator_readable': 'ArcA', 'function': 'Anaerobiosis; repression of aerobic growth genes during anaerobic growth', 'category': 'Uncharacterized', 'n_genes': 81, 'exp_var': 0.012209681631102, 'precision': 0.3333333333333333, 'recall': 0.1551724137931034, 'category_num': 74, '_key': 'iM92573382', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=74'}, {'name': 'yceO', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0006266741704267, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 75, '_key': 'iM26392615', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=75'}, {'name': 'Potassium', 'regulator_readable': 'KdpE', 'function': 'Potassium sensing and transport', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0007862495619258, 'precision': 0.6, 'recall': 1.0, 'category_num': 76, '_key': 'iM4524685', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=76'}, {'name': 'ROS TALE Del-1', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 14, 'exp_var': 0.0043178225193671, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 77, '_key': 'iM90583833', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=77'}, {'name': 'cyoB/kdpE/qseB KO', 'regulator_readable': 'NaN', 'function': 'cyoB/kdpE/qseF KO', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.0014471153027728, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 78, '_key': 'iM64557491', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=78'}, {'name': 'Isc System', 'regulator_readable': 'ryhB or IscR', 'function': 'Iron-sulfur cluster biogenesis and insertion', 'category': 'Uncharacterized', 'n_genes': 11, 'exp_var': 0.0017992544165329, 'precision': 0.7272727272727273, 'recall': 0.1311475409836065, 'category_num': 79, '_key': 'iM4758312', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=79'}, {'name': 'Thiamine-2', 'regulator_readable': 'NaN', 'function': 'Thiamine uptake and biosynthesis', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0022923918446863, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 80, '_key': 'iM90209158', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=80'}, {'name': 'crp KO-1', 'regulator_readable': 'NaN', 'function': 'crp KO and compensatory regulatory changes', 'category': 'Uncharacterized', 'n_genes': 24, 'exp_var': 0.004888264847324, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 81, '_key': 'iM17121236', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=81'}, {'name': 'SSW SNPs', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 89, 'exp_var': 0.0024338047760672, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 82, '_key': 'iM17380878', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=82'}, {'name': 'yafF', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0002556079055742, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 83, '_key': 'iM51697527', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=83'}, {'name': 'NtrC-2', 'regulator_readable': 'NtrC and RpoN', 'function': 'Nitrogen assimilation, ammonium transport', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.0031823700618902, 'precision': 0.8, 'recall': 0.1632653061224489, 'category_num': 84, '_key': 'iM21573578', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=84'}, {'name': 'DNA Damage', 'regulator_readable': 'LexA', 'function': 'SOS response to DNA damage or inhibition of DNA replication', 'category': 'Uncharacterized', 'n_genes': 18, 'exp_var': 0.0021546945665665, 'precision': 0.9444444444444444, 'recall': 0.293103448275862, 'category_num': 85, '_key': 'iM37913702', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=85'}, {'name': 'UC-3', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 51, 'exp_var': 0.0076599002079524, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 86, '_key': 'iM92983723', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=86'}, {'name': 'YieP', 'regulator_readable': 'YieP', 'function': 'Cellular membrane synthesis and stress response, 3-hydroxypropanoate tolerance', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.0010663800980329, 'precision': 0.6, 'recall': 1.0, 'category_num': 87, '_key': 'iM28653016', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=87'}, {'name': 'Nitrogen', 'regulator_readable': 'Nac', 'function': 'Nitrogen metabolism under nitrogen limitation', 'category': 'Uncharacterized', 'n_genes': 32, 'exp_var': 0.0015505296315246, 'precision': 0.4375, 'recall': 0.5384615384615384, 'category_num': 88, '_key': 'iM77585766', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=88'}, {'name': 'UC-6', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 11, 'exp_var': 0.0020332855180575, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 89, '_key': 'iM44458999', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=89'}, {'name': 'Leu/Val/Ile', 'regulator_readable': 'Lrp and IHF and Ile-tRNA and Leu-tRNA and Val-tRNA and ppGpp', 'function': 'Branched-chain amino acid biosynthesis', 'category': 'Uncharacterized', 'n_genes': 11, 'exp_var': 0.0020705618398104, 'precision': 0.4545454545454545, 'recall': 1.0, 'category_num': 90, '_key': 'iM80271701', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=90'}, {'name': 'NDH-1', 'regulator_readable': 'NarL and ArcA and IHF and Fnr and Fis and ryhB', 'function': 'NADH:ubiquinone oxidoreductase I: electron transfer from NADH to quinone pool, proton electrochemical gradient generation for aerobic and anaerobic respiration', 'category': 'Uncharacterized', 'n_genes': 12, 'exp_var': 0.001713813004514, 'precision': 0.75, 'recall': 0.6923076923076923, 'category_num': 91, '_key': 'iM96052223', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=91'}, {'name': 'Crp-1', 'regulator_readable': 'Crp', 'function': 'Miscellaneous carbon source metabolism', 'category': 'Uncharacterized', 'n_genes': 61, 'exp_var': 0.0104495830096349, 'precision': 0.3442622950819672, 'recall': 0.0405405405405405, 'category_num': 92, '_key': 'iM46776895', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=92'}, {'name': 'yneP', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.000408912393398, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 93, '_key': 'iM17035866', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=93'}, {'name': 'NtrC-1', 'regulator_readable': 'RpoN and RpoS and NtrC', 'function': 'Nitrogen assimilation, ammonium transport', 'category': 'Uncharacterized', 'n_genes': 18, 'exp_var': 0.0051747455543221, 'precision': 0.6666666666666666, 'recall': 0.7058823529411765, 'category_num': 94, '_key': 'iM79148494', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=94'}, {'name': 'Ribose', 'regulator_readable': 'dsrA and Crp and RbsR', 'function': 'Ribose transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0021421694670278, 'precision': 1.0, 'recall': 0.8333333333333334, 'category_num': 95, '_key': 'iM85847682', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=95'}, {'name': 'Dipeptide', 'regulator_readable': 'gcvB and IHF and Fnr', 'function': 'Dipeptide transport; flagellar biosynthesis', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.0036584076019571, 'precision': 0.625, 'recall': 0.8333333333333334, 'category_num': 96, '_key': 'iM81093080', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=96'}, {'name': 'Curli-1', 'regulator_readable': 'RpoS and FliZ and CpxR and CsgD and BtsR', 'function': 'Curli secretion and assembly', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0038668913830923, 'precision': 1.0, 'recall': 0.8571428571428571, 'category_num': 97, '_key': 'iM88948350', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=97'}, {'name': 'ymiB', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0003781235815983, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 98, '_key': 'iM95125414', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=98'}, {'name': 'yjjZ', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 2, 'exp_var': 0.0026166240662165, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 99, '_key': 'iM72013069', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=99'}, {'name': 'Molybdenum', 'regulator_readable': 'Crp and FliA and ModE', 'function': 'Molybdenum transport and molybdoenzyme synthesis', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0007605502780119, 'precision': 0.6, 'recall': 1.0, 'category_num': 100, '_key': 'iM58019', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=100'}, {'name': 'e14 Del', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 14, 'exp_var': 0.0028555173932445, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 101, '_key': 'iM3829852', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=101'}, {'name': 'Thr/Ser-1', 'regulator_readable': 'TdcA and TdcR and IHF and Fnr and Crp', 'function': 'Threonine/serine transport and catabolism during anaerobiosis', 'category': 'Uncharacterized', 'n_genes': 16, 'exp_var': 0.0040835359613318, 'precision': 0.4375, 'recall': 1.0, 'category_num': 102, '_key': 'iM77398177', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=102'}, {'name': 'Tryptophan', 'regulator_readable': 'TrpR and Trp-tRNA', 'function': 'Tryptophan transport and biosynthesis', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.0015520657370983, 'precision': 0.625, 'recall': 1.0, 'category_num': 103, '_key': 'iM92547337', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=103'}, {'name': 'FHL', 'regulator_readable': 'RpoN and NsrR and ModE and IHF and FhlA', 'function': 'Formate hydrogenlyase: hydrogen production during glucose fermentation via formate oxidation/proton reduction', 'category': 'Uncharacterized', 'n_genes': 7, 'exp_var': 0.0010700092076865, 'precision': 0.8571428571428571, 'recall': 0.6666666666666666, 'category_num': 104, '_key': 'iM33471869', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=104'}, {'name': 'Glycine', 'regulator_readable': 'PurR and Fnr and ppGpp and Crp and Lrp and GcvA', 'function': 'Glycine cleavage system', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0016625336133253, 'precision': 0.6, 'recall': 1.0, 'category_num': 105, '_key': 'iM52698747', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=105'}, {'name': 'BW25113', 'regulator_readable': 'NaN', 'function': 'Accounts for genomic differences between BW25113 and MG1655 substrains', 'category': 'Uncharacterized', 'n_genes': 26, 'exp_var': 0.0029903231929871, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 106, '_key': 'iM72027708', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=106'}, {'name': 'ypjJ', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 7, 'exp_var': 0.0003849885438514, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 107, '_key': 'iM85431971', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=107'}, {'name': 'PaaX', 'regulator_readable': 'Crp and IHF and PaaX and SlyA', 'function': 'Phenylacetic acid transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.001050651726133, 'precision': 0.7777777777777778, 'recall': 0.6363636363636364, 'category_num': 108, '_key': 'iM96008927', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=108'}, {'name': 'Flagella', 'regulator_readable': 'NaN', 'function': 'Flagellar biosynthesis', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0011778642609294, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 109, '_key': 'iM38135393', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=109'}, {'name': 'SCFA', 'regulator_readable': 'RpoN and AtoC and IHF', 'function': 'Short-chain fatty acid (acetoacetate) transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0004484520875545, 'precision': 1.0, 'recall': 1.0, 'category_num': 110, '_key': 'iM69730230', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=110'}, {'name': 'YgeV', 'regulator_readable': 'YgeV', 'function': 'Nucleoside degradation activated by ethanol treatment', 'category': 'Uncharacterized', 'n_genes': 13, 'exp_var': 0.0016177341446497, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 111, '_key': 'iM96221329', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=111'}, {'name': 'pts ALE', 'regulator_readable': 'NaN', 'function': 'ptsHIcrr KO and associated adaptive regulatory changes', 'category': 'Uncharacterized', 'n_genes': 37, 'exp_var': 0.0025699941179464, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 112, '_key': 'iM44463155', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=112'}, {'name': 'FDH-O', 'regulator_readable': 'SdhX', 'function': 'Formate dehydrogenase O; formate to carbon dioxide oxidation; formate-nitrate electron transport during aerobic/anerobic shift', 'category': 'Uncharacterized', 'n_genes': 7, 'exp_var': 0.0015333421694059, 'precision': 0.4285714285714285, 'recall': 0.5, 'category_num': 113, '_key': 'iM87217141', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=113'}, {'name': 'GlcNAc', 'regulator_readable': 'NagC and Crp', 'function': 'Amino sugar transport, biosynthesis and catabolism', 'category': 'Uncharacterized', 'n_genes': 17, 'exp_var': 0.0014408278781616, 'precision': 0.5294117647058824, 'recall': 0.391304347826087, 'category_num': 114, '_key': 'iM96109624', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=114'}, {'name': 'YgbI', 'regulator_readable': 'YgbI', 'function': 'Tartrate transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.000673002445767, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 115, '_key': 'iM81960360', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=115'}, {'name': 'Lysine/T2SS', 'regulator_readable': 'ArgP', 'function': 'Lysine transport and biosynthesis; type II secretion system', 'category': 'Uncharacterized', 'n_genes': 27, 'exp_var': 0.0023349720523286, 'precision': 0.2222222222222222, 'recall': 0.4615384615384615, 'category_num': 116, '_key': 'iM85148285', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=116'}, {'name': 'Fnr-3', 'regulator_readable': 'Fnr', 'function': 'Anaerobic growth', 'category': 'Uncharacterized', 'n_genes': 33, 'exp_var': 0.0149708875703917, 'precision': 0.3636363636363636, 'recall': 0.1188118811881188, 'category_num': 117, '_key': 'iM64622434', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=117'}, {'name': 'ythA', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 3, 'exp_var': 0.0004277535043072, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 118, '_key': 'iM68500007', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=118'}, {'name': 'entC/ubiC KO-1', 'regulator_readable': 'NaN', 'function': 'entC/ubiC KO and compensatory sad up-regulation', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0005228347399867, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 119, '_key': 'iM1913064', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=119'}, {'name': 'yoaI', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0010949437208195, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 120, '_key': 'iM78760678', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=120'}, {'name': 'Gluconate', 'regulator_readable': 'GntR and Crp', 'function': 'Gluconate transport and catabolism (Entner-Doudoroff)', 'category': 'Uncharacterized', 'n_genes': 12, 'exp_var': 0.0009070959842663, 'precision': 0.6666666666666666, 'recall': 0.8, 'category_num': 121, '_key': 'iM22309173', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=121'}, {'name': 'UC-9', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.0013455507887339, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 122, '_key': 'iM79972633', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=122'}, {'name': 'Copper', 'regulator_readable': 'CusR or CueR or HprR', 'function': 'Copper/silver sensing and export', 'category': 'Uncharacterized', 'n_genes': 11, 'exp_var': 0.0102481375207906, 'precision': 0.7272727272727273, 'recall': 0.8888888888888888, 'category_num': 123, '_key': 'iM18194462', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=123'}, {'name': 'Thr/Ser-2', 'regulator_readable': 'Fnr and Crp and IHF and TdcA and TdcR', 'function': 'Threonine/serine transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0009613941908135, 'precision': 0.8, 'recall': 0.5714285714285714, 'category_num': 124, '_key': 'iM8140367', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=124'}, {'name': 'Membrane', 'regulator_readable': 'NaN', 'function': 'Membrane-related stress response regulators, transporters, multidrug resistance', 'category': 'Uncharacterized', 'n_genes': 33, 'exp_var': 0.0053105853658499, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 125, '_key': 'iM1334654', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=125'}, {'name': 'Thiamine-1', 'regulator_readable': 'Thiamine diphosphate', 'function': 'Thiamine uptake and biosynthesis', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0090235477914814, 'precision': 0.8888888888888888, 'recall': 1.0, 'category_num': 126, '_key': 'iM19961817', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=126'}, {'name': 'Cold Shock', 'regulator_readable': 'NaN', 'function': 'Cold shock proteins', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.021490431196018, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 127, '_key': 'iM86173756', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=127'}, {'name': 'FliA', 'regulator_readable': 'FliA', 'function': 'Sigma 28; flagellum biosynthesis and export, motility', 'category': 'Uncharacterized', 'n_genes': 37, 'exp_var': 0.0263334493423872, 'precision': 0.918918918918919, 'recall': 0.4657534246575342, 'category_num': 128, '_key': 'iM70009740', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=128'}, {'name': 'Pyruvate-1', 'regulator_readable': 'BtsR', 'function': 'Pyruvate sensing, transport, and catabolism under nutrient limitation', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.002393605442079, 'precision': 0.5, 'recall': 0.2, 'category_num': 129, '_key': 'iM4662232', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=129'}, {'name': 'ldrD', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0005414098251802, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 130, '_key': 'iM86224082', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=130'}, {'name': 'efeU Activation', 'regulator_readable': 'NaN', 'function': 'efeU re-activation and associated efeU ALE-related expression changes', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0023523712057207, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 131, '_key': 'iM26065134', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=131'}, {'name': 'rcsB KO', 'regulator_readable': 'NaN', 'function': 'rcsB KO and associated regulatory changes', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.002465816337173, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 132, '_key': 'iM70221472', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=132'}, {'name': 'malE OE', 'regulator_readable': 'NaN', 'function': 'Accounts for malE plasmid overexpression', 'category': 'Uncharacterized', 'n_genes': 3, 'exp_var': 0.0005595901907518, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 133, '_key': 'iM9888697', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=133'}, {'name': 'Crp-2', 'regulator_readable': 'Crp', 'function': 'Carbohydrate transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 21, 'exp_var': 0.0083836495470449, 'precision': 0.7142857142857143, 'recall': 0.0289575289575289, 'category_num': 134, '_key': 'iM36162688', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=134'}, {'name': 'Cysteine-1', 'regulator_readable': 'CysB', 'function': 'Cysteine biosynthesis, sulfate assimilation', 'category': 'Uncharacterized', 'n_genes': 32, 'exp_var': 0.0070738307703427, 'precision': 0.5, 'recall': 0.5161290322580645, 'category_num': 135, '_key': 'iM91136681', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=135'}, {'name': 'Pyruvate-2', 'regulator_readable': 'PyrR', 'function': 'Pyruvate sensing and transport', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0019167457799291, 'precision': 1.0, 'recall': 1.0, 'category_num': 136, '_key': 'iM73213685', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=136'}, {'name': 'Phosphate-2', 'regulator_readable': 'PhoB', 'function': 'Inorganic phosphate sensing, transport, and production', 'category': 'Uncharacterized', 'n_genes': 20, 'exp_var': 0.0009951419487455, 'precision': 0.9, 'recall': 0.3829787234042553, 'category_num': 137, '_key': 'iM85160225', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=137'}, {'name': 'Leucine', 'regulator_readable': 'LeuO and Leu-tRNA and ppGpp', 'function': 'Leucine biosynthesis', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0027964595025327, 'precision': 0.8, 'recall': 1.0, 'category_num': 138, '_key': 'iM1018142', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=138'}, {'name': 'Capsule', 'regulator_readable': 'RcsAB', 'function': 'Colanic acid capsule synthesis and transport; periplasmic proteins; envelope stress response', 'category': 'Uncharacterized', 'n_genes': 38, 'exp_var': 0.0052519686749724, 'precision': 0.2631578947368421, 'recall': 0.625, 'category_num': 139, '_key': 'iM61764481', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=139'}, {'name': 'ytiC', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 2, 'exp_var': 0.0018282133653715, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 140, '_key': 'iM93161566', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=140'}, {'name': 'ROS TALE Del-2', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0031742139997406, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 141, '_key': 'iM30159485', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=141'}, {'name': 'UC-5', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.0018656431601232, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 142, '_key': 'iM28488447', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=142'}, {'name': 'FlhDC-1', 'regulator_readable': 'FlhDC', 'function': 'Flagellum biosynthesis and export', 'category': 'Uncharacterized', 'n_genes': 32, 'exp_var': 0.0059388907593869, 'precision': 0.71875, 'recall': 0.5897435897435898, 'category_num': 143, '_key': 'iM65728894', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=143'}, {'name': 'Putrescine', 'regulator_readable': 'RpoS and ArcA and PuuR', 'function': 'Putrescine transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.0113710936848435, 'precision': 1.0, 'recall': 1.0, 'category_num': 144, '_key': 'iM60843721', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=144'}, {'name': 'ppGpp', 'regulator_readable': 'ppGpp', 'function': 'Stringent response', 'category': 'Uncharacterized', 'n_genes': 70, 'exp_var': 0.0232606932974963, 'precision': 0.9142857142857144, 'recall': 0.0846560846560846, 'category_num': 145, '_key': 'iM11186439', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=145'}, {'name': 'Glycolate', 'regulator_readable': 'IHF and ArcA and PdhR and GlcC', 'function': 'Glycolate/glyoxylate tranport and catabolism', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0058865397941735, 'precision': 1.0, 'recall': 1.0, 'category_num': 146, '_key': 'iM25796760', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=146'}, {'name': 'PAL 26 Del', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 43, 'exp_var': 0.0014386774606234, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 147, '_key': 'iM3492238', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=147'}, {'name': 'LPS', 'regulator_readable': 'NaN', 'function': 'Lipopolysaccharide biosynthesis and export', 'category': 'Uncharacterized', 'n_genes': 43, 'exp_var': 0.0059235570640736, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 148, '_key': 'iM2254006', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=148'}, {'name': 'Arginine', 'regulator_readable': 'ArgR', 'function': 'Arginine transport and biosynthesis', 'category': 'Uncharacterized', 'n_genes': 24, 'exp_var': 0.0075726204463577, 'precision': 0.8333333333333334, 'recall': 0.5405405405405406, 'category_num': 149, '_key': 'iM39351374', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=149'}, {'name': 'Nickel/Cobalt', 'regulator_readable': 'RcnR', 'function': 'Nickel and cobalt resistance via efflux', 'category': 'Uncharacterized', 'n_genes': 2, 'exp_var': 0.0012995110204339, 'precision': 1.0, 'recall': 0.6666666666666666, 'category_num': 150, '_key': 'iM84731036', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=150'}, {'name': 'IS5', 'regulator_readable': 'NaN', 'function': 'IS5 transcripts', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0027587036419692, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 151, '_key': 'iM76015627', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=151'}, {'name': 'NtrC-3', 'regulator_readable': 'NtrC and RpoN', 'function': 'Nitrogen assimilation, pyrimidine transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 20, 'exp_var': 0.0026376541759355, 'precision': 0.7, 'recall': 0.2857142857142857, 'category_num': 152, '_key': 'iM39158158', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=152'}, {'name': 'arcA/luxS KO', 'regulator_readable': 'NaN', 'function': 'arcA KO', 'category': 'Uncharacterized', 'n_genes': 2, 'exp_var': 0.0002828154638168, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 153, '_key': 'iM22861277', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=153'}, {'name': 'sdh KO', 'regulator_readable': 'NaN', 'function': 'sdh KO', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0029264852749815, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 154, '_key': 'iM72108024', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=154'}, {'name': 'YmfT', 'regulator_readable': 'YmfT', 'function': 'e14 prophage regulator activated by DNA damage', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.0010120029153843, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 155, '_key': 'iM20636618', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=155'}, {'name': 'Fnr-2', 'regulator_readable': 'Fnr', 'function': 'Transition from aerobic to anaerobic growth; global control of anaerobiosis', 'category': 'Uncharacterized', 'n_genes': 62, 'exp_var': 0.0095387962854195, 'precision': 0.3225806451612903, 'recall': 0.0675675675675675, 'category_num': 156, '_key': 'iM5438002', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=156'}, {'name': 'Microaerobic', 'regulator_readable': 'RpoS and Fis and NarP and NarL and AppY and YdeO and ArcA and IscR', 'function': 'Hydrogenase 1 and cytochrome bd-II ubiquinol oxidase: microaerobic respiratory chain between hydrogen and oxygen', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0026364308308872, 'precision': 1.0, 'recall': 1.0, 'category_num': 157, '_key': 'iM19787678', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=157'}, {'name': 'Cysteine-2', 'regulator_readable': 'CysB', 'function': 'Cysteine biosynthesis, sulfate assimilation', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0009643063423739, 'precision': 0.7777777777777778, 'recall': 0.2258064516129032, 'category_num': 158, '_key': 'iM71592840', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=158'}, {'name': 'ydfW', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 3, 'exp_var': 0.0002876525225196, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 159, '_key': 'iM87614300', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=159'}, {'name': 'YcjW', 'regulator_readable': 'YcjW', 'function': 'Carbohydrate transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 17, 'exp_var': 0.0005891164112311, 'precision': 0.7058823529411765, 'recall': 0.1643835616438356, 'category_num': 160, '_key': 'iM88093741', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=160'}, {'name': 'PAL Amp', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 88, 'exp_var': 0.001002999618594, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 161, '_key': 'iM43595205', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=161'}, {'name': 'Propionate', 'regulator_readable': 'PrpR and RpoN and Cra and Crp', 'function': 'Propionate catabolism; 2-methylcitric acid cycle', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0042154113788448, 'precision': 1.0, 'recall': 1.0, 'category_num': 162, '_key': 'iM83722955', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=162'}, {'name': 'CP4-44', 'regulator_readable': 'NaN', 'function': 'CP4-44 prophage expression variation', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0014264960630087, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 163, '_key': 'iM56163907', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=163'}, {'name': 'minicoli KOs', 'regulator_readable': 'NaN', 'function': 'Deletions from minicoli strain', 'category': 'Uncharacterized', 'n_genes': 115, 'exp_var': 0.0466260940264305, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 164, '_key': 'iM10685809', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=164'}, {'name': 'GadXW', 'regulator_readable': 'GadX and GadW', 'function': 'Glutamate-dependent acid resistance system; multidrug efflux', 'category': 'Uncharacterized', 'n_genes': 19, 'exp_var': 0.0181034365618988, 'precision': 0.5789473684210527, 'recall': 0.8461538461538461, 'category_num': 165, '_key': 'iM66358263', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=165'}, {'name': 'Rhamnose', 'regulator_readable': 'Crp and RhaS', 'function': 'Rhamnose transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.0049969961718729, 'precision': 0.75, 'recall': 1.0, 'category_num': 166, '_key': 'iM81159711', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=166'}, {'name': 'UC-2', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 96, 'exp_var': 0.0176945831137673, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 167, '_key': 'iM27280308', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=167'}, {'name': 'ydfB', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 1, 'exp_var': 0.0004215509568046, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 168, '_key': 'iM47309714', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=168'}, {'name': 'NrdR', 'regulator_readable': 'NrdR', 'function': 'Ribonucleoside reduction', 'category': 'Uncharacterized', 'n_genes': 7, 'exp_var': 0.0036557335307761, 'precision': 0.8571428571428571, 'recall': 0.6666666666666666, 'category_num': 169, '_key': 'iM36688207', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=169'}, {'name': 'Purine', 'regulator_readable': 'PurR and ppGpp', 'function': 'Purine transport and biosynthesis', 'category': 'Uncharacterized', 'n_genes': 22, 'exp_var': 0.0089683253261392, 'precision': 0.5454545454545454, 'recall': 0.5714285714285714, 'category_num': 170, '_key': 'iM3678597', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=170'}, {'name': 'cyoB/ndh/nuoB KO-1', 'regulator_readable': 'NaN', 'function': 'NDH-1 (nuoB) KO and compensatory NDH-2 (ndh) up-regulation', 'category': 'Uncharacterized', 'n_genes': 3, 'exp_var': 0.0006249431938155, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 171, '_key': 'iM56087303', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=171'}, {'name': 'Fimbriae', 'regulator_readable': 'QseB and IHF and Lrp and H-NS', 'function': 'Type 1 (mannose-sensitive) pili/fimbriae biosynthesis and secretion', 'category': 'Uncharacterized', 'n_genes': 8, 'exp_var': 0.004407357928021, 'precision': 0.875, 'recall': 1.0, 'category_num': 172, '_key': 'iM58325981', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=172'}, {'name': 'IS Elements-1', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 24, 'exp_var': 0.0040923738515132, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 173, '_key': 'iM28457569', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=173'}, {'name': 'Cra', 'regulator_readable': 'Cra', 'function': 'Central carbon metabolism; glycolytic flux-dependent regulation via fructose sensing', 'category': 'Uncharacterized', 'n_genes': 17, 'exp_var': 0.0039547447575098, 'precision': 0.7647058823529411, 'recall': 0.1666666666666666, 'category_num': 174, '_key': 'iM48712290', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=174'}, {'name': 'Hot TALE 16', 'regulator_readable': 'NaN', 'function': 'Regulatory changes under high temperature', 'category': 'Uncharacterized', 'n_genes': 17, 'exp_var': 0.0007551432881904, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 175, '_key': 'iM62499323', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=175'}, {'name': 'Suf System', 'regulator_readable': 'NsrR and IHF and IscR and OxyR and Fur and ppGpp', 'function': 'Iron-sulfur cluster biosynthesis and insertion', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0028037304268141, 'precision': 1.0, 'recall': 1.0, 'category_num': 176, '_key': 'iM85375437', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=176'}, {'name': 'PlaR', 'regulator_readable': 'PlaR and RpoE and Crp and IHF', 'function': 'Plant-derived sugars transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.0005738877582043, 'precision': 0.9, 'recall': 1.0, 'category_num': 177, '_key': 'iM6583188', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=177'}, {'name': 'NO Stress', 'regulator_readable': 'NsrR or NarP', 'function': 'Nitric oxide sensing and detoxification, iron-sulfur cluster repair, anaerobic growth', 'category': 'Uncharacterized', 'n_genes': 11, 'exp_var': 0.0008460152620453, 'precision': 0.8181818181818182, 'recall': 0.075, 'category_num': 178, '_key': 'iM26792835', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=178'}, {'name': 'Maltose', 'regulator_readable': 'MalT', 'function': 'Maltose transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0070724640931937, 'precision': 0.8888888888888888, 'recall': 1.0, 'category_num': 179, '_key': 'iM12161383', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=179'}, {'name': 'Tyr/Trp/Phe', 'regulator_readable': 'TyrR or TrpR or Trp-tRNA', 'function': 'Aromatic amino acid biosynthesis and transport', 'category': 'Uncharacterized', 'n_genes': 12, 'exp_var': 0.0017713137524119, 'precision': 0.9166666666666666, 'recall': 0.5789473684210527, 'category_num': 180, '_key': 'iM52700628', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=180'}, {'name': 'GadX', 'regulator_readable': 'RpoS and GadX', 'function': 'Glutamate-dependent acid resistance system; multidrug efflux', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0020686271304211, 'precision': 0.6666666666666666, 'recall': 0.375, 'category_num': 181, '_key': 'iM47164636', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=181'}, {'name': 'Fatty Acid', 'regulator_readable': 'FadR', 'function': 'Fatty acid degradation and biosynthesis', 'category': 'Uncharacterized', 'n_genes': 7, 'exp_var': 0.0026795974164881, 'precision': 1.0, 'recall': 0.3181818181818182, 'category_num': 182, '_key': 'iM24381865', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=182'}, {'name': 'DhaR', 'regulator_readable': 'DhaR', 'function': 'Dihydroxyacetone transport and catabolism; phosphotransferase activity for DhaM phosphorylation', 'category': 'Uncharacterized', 'n_genes': 13, 'exp_var': 0.0016691607085843, 'precision': 0.2307692307692307, 'recall': 0.75, 'category_num': 183, '_key': 'iM24587416', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=183'}, {'name': 'pts KO', 'regulator_readable': 'NaN', 'function': 'ptsHI/crr KO and compensatory regulatory changes', 'category': 'Uncharacterized', 'n_genes': 13, 'exp_var': 0.0009264405588177, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 184, '_key': 'iM46973986', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=184'}, {'name': 'gcvB', 'regulator_readable': 'gcvB', 'function': 'Amino acid transport', 'category': 'Uncharacterized', 'n_genes': 76, 'exp_var': 0.0055535628716726, 'precision': 0.1842105263157894, 'recall': 0.2456140350877192, 'category_num': 185, '_key': 'iM37128414', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=185'}, {'name': 'Glutarate', 'regulator_readable': 'RpoS and H-NS and Crp and Lrp and GlaR and ppGpp', 'function': 'Glutarate/aminobutyrate transport and catabolism', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0031702833722321, 'precision': 1.0, 'recall': 1.0, 'category_num': 186, '_key': 'iM72407541', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=186'}, {'name': 'entC/ubiC KO-2', 'regulator_readable': 'NaN', 'function': 'entC/ubiC KO and compensatory sad up-regulation', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0007308965174548, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 187, '_key': 'iM69451395', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=187'}, {'name': 'Oxidoreductase', 'regulator_readable': 'NaN', 'function': 'Putative oxidoreductase components', 'category': 'Uncharacterized', 'n_genes': 5, 'exp_var': 0.0006885227037741, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 188, '_key': 'iM65593873', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=188'}, {'name': 'UC-7', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 32, 'exp_var': 0.0125554454651567, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 189, '_key': 'iM38241260', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=189'}, {'name': 'Efflux Pump', 'regulator_readable': 'NaN', 'function': 'Multidrug/spermidine efflux pump', 'category': 'Uncharacterized', 'n_genes': 2, 'exp_var': 0.0019208357521413, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 190, '_key': 'iM59911901', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=190'}, {'name': 'Tryptophanase', 'regulator_readable': 'Crp and TorR and Nac and L-tryptophan', 'function': 'Tryptophan import and catabolism via tryptophanase; indole/pyruvate biosynthesis', 'category': 'Uncharacterized', 'n_genes': 2, 'exp_var': 0.0032483463491277, 'precision': 1.0, 'recall': 1.0, 'category_num': 191, '_key': 'iM99512115', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=191'}, {'name': 'crp KO-2', 'regulator_readable': 'NaN', 'function': 'crp KO and associated glycerol-3-phosphate (GlpR) regulatory changes', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0006036042182727, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 192, '_key': 'iM11264473', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=192'}, {'name': 'cyoB/ndh/nuoB KO-2', 'regulator_readable': 'NaN', 'function': 'cyoB/ndh/nuoB KOs', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0010265287046586, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 193, '_key': 'iM58701030', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=193'}, {'name': 'tpiA KO', 'regulator_readable': 'NaN', 'function': 'tpiA KO and compensatory evolved regulatory changes', 'category': 'Uncharacterized', 'n_genes': 10, 'exp_var': 0.0005477001860416, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 194, '_key': 'iM18447210', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=194'}, {'name': 'Phage Shock', 'regulator_readable': 'RpoN and IHF and PspF', 'function': 'Phage shock response; response to filamentous phage infection and/or extracytoplasmic shock', 'category': 'Uncharacterized', 'n_genes': 6, 'exp_var': 0.0036314257008215, 'precision': 0.8333333333333334, 'recall': 0.8333333333333334, 'category_num': 195, '_key': 'iM39099540', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=195'}, {'name': 'Glyoxylate', 'regulator_readable': 'Crp and ArcA and ppGpp and Cra and IclR and IHF', 'function': 'Glyxoylate cycle; alternate to TCA cycle', 'category': 'Uncharacterized', 'n_genes': 3, 'exp_var': 0.0020848833793986, 'precision': 1.0, 'recall': 1.0, 'category_num': 196, '_key': 'iM30607027', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=196'}, {'name': 'CCK-PGI Del', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 15, 'exp_var': 0.0007447751672728, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 197, '_key': 'iM73656463', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=197'}, {'name': 'Acetate', 'regulator_readable': 'NaN', 'function': 'Acetate uptake and catabolism; aldehyde catabolism; acetate/ethanol detoxification at stationary phase', 'category': 'Uncharacterized', 'n_genes': 7, 'exp_var': 0.0061907381633782, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 198, '_key': 'iM93156217', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=198'}, {'name': 'Ile/Val', 'regulator_readable': 'NaN', 'function': 'Isoleucine and valine biosynthesis', 'category': 'Uncharacterized', 'n_genes': 4, 'exp_var': 0.0010815895254098, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 199, '_key': 'iM52820144', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=199'}, {'name': 'UC-8', 'regulator_readable': 'NaN', 'function': 'Unknown', 'category': 'Uncharacterized', 'n_genes': 9, 'exp_var': 0.0083955528371049, 'precision': 'NaN', 'recall': 'NaN', 'category_num': 200, '_key': 'iM3163427', 'linkout': 'https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1k&k=200'}]\n"
     ]
    }
   ],
   "source": [
    "# Generate iModulon collection\n",
    "collections['iModulon'] = []\n",
    "for organism in imodulon_files:\n",
    "    for dataset in imodulon_files[organism]:\n",
    "        imodulon_path = imodulon_files[organism][dataset]\n",
    "        collections['iModulon'].extend(get_imodulon_collection(organism=organism, dataset=dataset, table_path=imodulon_path))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb9ac80f",
   "metadata": {},
   "source": [
    "### 7. Output nodes and edges collections as two JSON files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "c6c083fa-6595-47fa-98ce-8c9e6303ced8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output nodes collections \n",
    "\n",
    "os.makedirs(os.path.dirname(nodes_path), exist_ok = True)\n",
    "nodes_str = json.dumps(collections)\n",
    "with open(nodes_path, 'w') as out:\n",
    "    out.write(nodes_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "cc77c8fc-1de8-4d62-aa43-6e2671b3d459",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output edges collections\n",
    "os.makedirs(os.path.dirname(edges_path), exist_ok = True)\n",
    "edges_str = json.dumps(edges)\n",
    "with open(edges_path, 'w') as out:\n",
    "    out.write(edges_str)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
